pkg/workceptor/kubernetes.go

```diff
@@ -338,115 +343,174 @@ func (kw *KubeUnit) KubeLoggingWithReconnect(streamWait *sync.WaitGroup, stdout
+					// Their are many reasons why the kube api might not be able to get the pod,
+					// This does not mean their is a problem just yet.
+					// Lets try to get the pod again, max 5 times, and decide.
+					kw.GetWorkceptor().nc.GetLogger().Debug("Error getting pod after reading stream: '%s' , continuing try to get pod up to 5 more times.", kubeErr)
+
+					continue mainLoop
 				}
-				for _, condition := range erroredPod.Status.Conditions {
-					if condition.Type == corev1.PodReady && condition.Status == corev1.ConditionTrue {
-						podConditionReady = true
+
+				var containerState corev1.ContainerState
+				foundContainer := false
+				for _, containerStatus := range podDetails.Status.ContainerStatuses {
+					if containerStatus.Name == WorkerContainerName {
+						containerState = containerStatus.State
+						foundContainer = true
 					}
 				}
 
-				if err == io.EOF && !podConditionReady {
+				if !foundContainer {
+					kw.GetWorkceptor().nc.GetLogger().Error("Unable to find the container %s for pod %s. This is unrecoverable. Marking the job as failed and exiting", WorkerContainerName, podName)
+					*stdoutErr = fmt.Errorf("unable to find the container %s for pod %s. This is unrecoverable. Marking the job as failed and exiting", WorkerContainerName, podName)
+
+					return
+				}
+
+				switch {
+				case containerState.Running != nil:
+					// We got EOF but pod is running, is this because we checked too fast? Will it turn into a terminated state soon or are we hitting the 4 hour log stream kube error. We will attempt to reconnect a max of 5 times in order to cover both cases
+					// If we can't get reconnect without an EOF we will error and mark the job as failed.
+					retryGetLogStream--
+					if retryGetLogStream > 0 {
+						kw.GetWorkceptor().nc.GetLogger().Info(
+							"Detected Error: %s for pod %s/%s. Will retry %d more times.",
+							err,
+							podNamespace,
+							podName,
+							retryGetLogStream,
+						)
+
+						time.Sleep(time.Second * time.Duration(curContainerDelay))
+						prevContainerDelay, curContainerDelay = curContainerDelay, prevContainerDelay+curContainerDelay
+
+						continue mainLoop
+					}
+					// Retrying hasn't worked we will error and mark the job as failed
+					kw.GetWorkceptor().nc.GetLogger().Debug("Container in %s pod is running but unable to attach to the log stream", WorkerContainerName)
+					*stdoutErr = fmt.Errorf("detected Error: %s for pod %s/%s. Pod is running but unable to attach to the log stream", err,
+						podNamespace,
+						podName,
+					)
+
+					return
+				case containerState.Terminated != nil:
+					// We got EOF and the pod terminated in a bad state, we will return the error and mark the job as failed.
+					if containerState.Terminated.ExitCode != 0 {
+						kw.GetWorkceptor().nc.GetLogger().Debug("Container in %s pod has terminated, with terminated exit code: %v", WorkerContainerName, containerState.Terminated.ExitCode)
```

Comment:
Do we want `Debug()` here?
```suggestion
						kw.GetWorkceptor().nc.GetLogger().Info("Container in %s pod has terminated, with nonzero exit code: %v", WorkerContainerName, containerState.Terminated.ExitCode)
```

Minor suggestion - In this case I think receptor did its job so I would log it as Info (not Error), but I think putting a word like `nonzero` will make it easier to search for these.